Layer documentation enhancements (#125)
napari/docs: docs/howtos/layers/labels.md

  will not be affected. It looks like this: ![image: eraser tool](../../images/labels-layer-eraser.png)

* **Paintbrush**

  One of the major use cases for the `labels layer` is to manually edit or
  create image segmentations. One of the tools that can be used for manual
  editing is the `paintbrush`, activated by clicking the `paintbrush` icon in
  the `layer controls` panel. Once the paintbrush is enabled, the pan and zoom
  functionality of the  viewer canvas is disabled, and you can paint on the
  canvas. You can temporarily re-enable pan and zoom by pressing and holding the
  spacebar. This feature is useful if you want to move around the `labels layer`
  as you paint.

  Click the `paintbrush` icon and select a color from the `label` option by
  clicking on the + or - on the label bar in the layer controls panel. This will
  scroll through the available colors.  Whatever color you pick will be the
  *edge color* of the label. Draw the edge of the label using the `paintbrush`.
  If you draw a continuous edge, you can fill it in using the `paint bucket` or
  `fill bucket` tool. It can be the same color as the edge or a different color.

>   Adjust the size of your `paintbrush` using the `brush size` slider. It can be

Reviewer: ```suggestion
  Adjust the size of your `paintbrush` using the `brush size` slider or using
  the default keybindings: `[` and `]`. It can be
```

Reviewer 2: I made this change in codespace.

* Brush size

  Adjust the size of the `paintbrush` using the `brush size` slider to any value
  from 1 to 40. 1 is as small as a single pixel.

* Blending

  Select from `translucent`, `translucent no depth`, `additive`, `minimum`, or
  `opaque` from the dropdown. Refer to the [Blending layers](blending-layers)
  section of _Layers at a glance_ for an explanation of each type of blending.

* Color mode

  Select `auto` or `direct` from the dropdown. Auto is the default and allows
  color to be set via a hash function with a seed. Direct allows the color of
  each label to be set directly by a color dictionary.

* Contour

  If this combo box has any value other than 0, only the contours of the labels
  will show. Change the value by clicking the - or + symbols on either end of
  the bar.

* n edit dim

  This is the number of dimensions across which labels will be edited.

* Contiguous

  If this box is checked the `fill bucket` changes only connected pixels of the
  same label.

* Preserve labels

  If this box is checked, existing labels are preserved while painting. It
  defaults to false to allow painting on existing labels. When set to true,

>   existing labels will be preserved during painting.

Reviewer: ```suggestion
  existing labels will be preserved during painting. You can toggle
  this mode using the default keybinding `p`.
```

* Show selected

  When this is checked only the selected labels will be displayed. Selected
  labels are those labels that match the color in the `label` combo box. When it
  is not checked, all labels will be displayed.

## Editing using the tools in the GUI
### Pan and zoom mode

The default mode of the `labels layer` is to support panning and zooming. This
mode is represented by the magnifying glass in the `layer controls` panel. While
pan and zoom is selected editing the layer is not possible. Once you click on
one of the editing tools, pan and zoom is turned off. Return to pan and zoom
mode by pressing the `4` key when the `labels layer` is selected.

### Creating a new `labels layer`

Create a brand-new empty `labels layer` by clicking the `New labels layer`
button above the layer list. The shape of the new labels layer will match the
size of any currently existing image layers, allowing you to paint on top of
them.

### Selecting a label

A particular label can be chosen in one of three ways:

* Using the label combobox inside the `layer controls` panel and typing in the
  numeric value of the desired label 

> * Using the + or - buttons to get to the desired label color

Reviewer: ```suggestion
* Using the + or - buttons to get to the desired label color or press the default 
keybinding `m` to set a new label.
```

* Selecting the `color picker` tool and then clicking on a pixel with the
  desired label color in the image

When a label is chosen its integer appears inside the label combobox and the
color of the label is shown in the thumbnail next to the label combobox. If the
0 label is selected, then a checkerboard pattern is shown in the thumbnail to
represent the transparent color.

You can quickly select the color picker by pressing the `4` key when the labels
layer is selected.

You can set the selected label to 0, the background label, by pressing `1`.

You can set the selected label to be one larger than the current largest label

> by pressing `m` with either the `paintbrush` or `fill bucket` selected. This

Reviewer: ```suggestion
You can set the selected label to a new label—one larger than the current 
largest label—by pressing `m` with either the `paintbrush` or `fill bucket` selected. This
```

Reviewer 2: It literally adds 1 to the intger that represents the label color.

Reviewer 2: I made this change in codespace.

selection will guarantee that you are using a label that hasn't been used before.

<!-- I couldn't get this option to work.
"You can also increment or decrement the currently selected label by pressing the

> `I` or `D` key, respectively." -->

Reviewer: ```suggestion
You can also increment or decrement the currently selected label by pressing (by default) the
`=` or `-` key, respectively.
```
Updated for the latest default keybinds.

Reviewer 2: I'm not sure why we would add (by default)?

Reviewer 2: I want to change this manually to, "You can also increment or decrement the currently selected label by pressing the default keybinding `=` or `-` keys on the keyboard."

### Creating, deleting, merging, and splitting connected components

Create and edit object segmentation maps using the `color picker`, `paintbrush`,
and `fill bucket` tools. Below we show how to use these tools by performing
common editing tasks on connected components (keep the `contiguous` box checked).

* Creating or drawing a connected component
  ![image: draw component](../../images/draw_component.webm)
  * Press `m` to select a label color that has not been used.
  * Select the `paintbrush` tool and draw a closed contour around the object.
  * Select the `fill bucket` tool and click inside the contour to assign the
    label to all pixels of the object.

* Deleting a connected component
  ![image: delete label](../../images/delete_label.webm)
  Select the background label with the `color picker` (alternative: press

>   keyboard shortcut `E`), then use the `fill bucket` to set all pixels of the

Reviewer: ```suggestion
  Select the background label with the `color picker`, then use the 
  `fill bucket` to set all pixels of the
```
E doesn't appear to do that anymore and I don't see such an option in the settings.
Will make issue on napari/napari to add (back?) this feature.

Reviewer: https://github.com/napari/napari/issues/5666

Reviewer: PR: https://github.com/napari/napari/pull/5672

Reviewer 2: I believe this actually  works but I will test it again.  First you have to set up connected components, etc.  (Nadalyn - look at this.)

  connected component to background.

* Merging connected components
  ![image: merge labels](../../images/merge_labels.webm)
  * Select the label of one of the components with the `color picker` tool.
  * Select the `fill bucket` and fill the components to be merged.

* Splitting a connected component
  ![image: split label](../../images/split_label.webm)  
  Splitting a connected component will introduce an additional object.
  * Press `m` to select a label not already in use.
  * Use the `paintbrush` tool to draw a dividing line where you want to split
    the component.
  * Assign the new label to one of the parts with the `fill bucket`.

### Undo/redo functionality

When using the `fill bucket` or `paintbrush` it can be easy to make a mistake
that you might want to undo or you might want to redo something that has just
been undone. Use  `ctrl-z` to redo and `shift-ctrl-z` to redo. There are plans
to support this sort of functionality more generally, but for now these actions
will undo the most recent painting or filling event, up to 100 events in the
past.

If you have multidimensional data, adjusting the currently viewed slice will

> cause the undo history to be reset.

Reviewer: This is a big deal, maybe we should make this a more prominent warning/note?

Reviewer 2: Yes.  You're right.  I'll get someone to show me how to create an admonition.

mode by pressing the `4` key when the `labels layer` is selected.

### Creating a new `labels layer`

Create a brand-new empty `labels layer` by clicking the `New labels layer`
button above the layer list. The shape of the new labels layer will match the
size of any currently existing image layers, allowing you to paint on top of
them.

### Selecting a label

A particular label can be chosen in one of three ways:

* Using the label combobox inside the `layer controls` panel and typing in the
  numeric value of the desired label 

* Using the + or - buttons to get to the desired label color or press the default 
keybinding `m` to set a new label.

* Selecting the `color picker` tool and then clicking on a pixel with the
  desired label color in the image

When a label is chosen its integer appears inside the label combobox and the
color of the label is shown in the thumbnail next to the label combobox. If the
0 label is selected, then a checkerboard pattern is shown in the thumbnail to
represent the transparent color.

You can quickly select the color picker by pressing the `4` key when the labels
layer is selected.

> You can set the selected label to 0, the background label, by pressing `b`.

Author: Fix this! @psobolewskiPhD

> You can set the selected label to 0, the background label, by pressing `b`.

Reviewer: ```suggestion
While painting with a label, you can swap the current (selected) label to the transparent background label (`0`) by pressing `x`. Pressing `x` again will restore the last used label.
```
How does this sound? Trying to match the wording above.

Reviewer: Another simpler option
> While painting with a label, you can toggle between the current (selected) label and the transparent background label (`0`) by pressing `x`.

or

> While painting with a label, you can swap between the current (selected) label and the transparent background label (`0`) by pressing `x`.

Author: Accepted your third option!

label control and the color of the label is shown in the thumbnail next to the
control. If the 0 label is selected, then a checkerboard pattern is shown in the
thumbnail to represent the transparent color.

You can quickly select the color picker by pressing the `5` key when the labels
layer is selected.

While painting with a label, you can swap between the current (selected) label
and the transparent background label (`0`) by pressing `x`.

You can set the selected label to a new label -- one larger than the current
largest label -- by pressing `m` with either the `paintbrush` or `fill bucket`
tools selected. This selection will guarantee that you are using a label that
hasn't been used before.

You can also increment or decrement the currently selected label by pressing the
`=` or `-` keys, respectively.

### Creating, deleting, merging, and splitting connected components

Create and edit object segmentation maps using the `color picker`, `paintbrush`,
and `fill bucket` tools. Below we show how to use these tools by performing
common editing tasks on connected components (keep the `contiguous` box checked).

* Creating or drawing a connected component

>   ![Using the paintbrush and fill bucket tools to draw a connected component](../../images/draw_component.webm)

Reviewer: This video has sound 🤣 
You were watching some TV while doing these!

  * Press `m` to select a label color that has not been used.
  * Select the `paintbrush` tool and draw a closed contour around the object.
  * Select the `fill bucket` tool and click inside the contour to assign the
    label to all pixels of the object.

* Deleting a connected component

>   ![Deleting selected labels](../../images/delete_label.webm)

Reviewer: This video has sound 🤣

>   Select the background label with the `color picker`, then use the 

Reviewer: ```suggestion
  Select the background label with the `color picker` or press `x`, then use the 
```

  `fill bucket` to set all pixels of the
  connected component to background.

* Merging connected components

>   ![Selecting a label and merging with a connecting label](../../images/merge_labels.webm)

Reviewer: This video has sound 🤣

  * Select the label of one of the components with the `color picker` tool.
  * Select the `fill bucket` and fill the components to be merged.

* Splitting a connected component

>   ![Using the paintbrush tool to split a label into two](../../images/split_label.webm)

Reviewer: This video has sound 🤣

  Splitting a connected component will introduce an additional object.

>   * Press `m` to select a label not already in use.

Reviewer: ```suggestion
  * Select the background label with the `color picker` or press `x`.
```
Using `m` in this case would paint the line with a new label, adding a component.

viewer.close()
```

### Arguments of `view_labels` and `add_labels`

{meth}`~napari.view_layers.view_labels` and {meth}`~napari.Viewer.add_labels`
accept the same layer-creation parameters.

```{code-cell} python
:tags: [hide-output]

help(napari.view_labels)

> ```

Reviewer: This code cell seems wierd? Here's what I see:
<img width="764" alt="image" src="https://user-images.githubusercontent.com/76622105/227726638-930c071c-eddf-472b-82f9-7b0b81ca1ef6.png">

I'm not sure that's useful alone like that? are we telling the user to type that to see the documentation?

Author: Yes - and if you click the arrow it will show you the full output of the help on napari.view_labels. It was originally fully hidden which I think is even less useful 😅 But we can explore other options if you have suggestions

Reviewer: I think i would either:
1) show output and say here's the blah blah
or 
2) say something like `you can see the full blah blah by using the following command`

Reviewer 2: Good point.

Reviewer 2: I made this change in codespace.
